samza-sql/src/main/java/org/apache/samza/sql/runner/SamzaSqlApplicationRunner.java

```diff
@@ -0,0 +1,134 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.apache.samza.sql.runner;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.lang3.Validate;
+import org.apache.samza.application.StreamApplication;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.MapConfig;
+import org.apache.samza.job.ApplicationStatus;
+import org.apache.samza.runtime.AbstractApplicationRunner;
+import org.apache.samza.runtime.ApplicationRunner;
+import org.apache.samza.runtime.LocalApplicationRunner;
+import org.apache.samza.runtime.RemoteApplicationRunner;
+import org.apache.samza.sql.interfaces.SourceResolver;
+import org.apache.samza.sql.interfaces.SystemStreamConfig;
+import org.apache.samza.sql.testutil.SamzaSqlQueryParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Application runner implementation for SamzaSqlApplication.
+ * SamzaSqlApplication needs SamzaSqlConfigRewriter to infer some of the configs from SQL statements.
+ * Since Samza's config rewriting capability is available only in the RemoteApplicationRunner.
+ * This runner invokes the SamzaSqlConfig re-writer if it is invoked on a standalone mode (i.e. localRunner == true)
+ * otherwise directly calls the RemoteApplicationRunner which automatically performs the config rewriting .
+ */
+public class SamzaSqlApplicationRunner extends AbstractApplicationRunner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SamzaSqlApplicationRunner.class);
+
+  private final Config sqlConfig;
+  private final ApplicationRunner appRunner;
+  private final Boolean localRunner;
+
+  public static final String RUNNER_CONFIG = "app.runner.class";
+  public static final String CFG_FMT_SAMZA_STREAM_SYSTEM = "streams.%s.samza.system";
+
+  public SamzaSqlApplicationRunner(Config config) {
+    this(false, config);
+  }
+
+  public SamzaSqlApplicationRunner(Boolean localRunner, Config config) {
+    super(config);
+    this.localRunner = localRunner;
+    sqlConfig = computeSamzaConfigs(localRunner, config);
+    appRunner = ApplicationRunner.fromConfig(sqlConfig);
```

Comment:
Cann't we derive the boolean value localRunner from config? We don't have many ApplicationRunner classes implemented. And if needed, we can add an interface method isLocalRunner() to ApplicationRunner interface, instead of asking user to supply a separate boolean variable in the constructor.

Reply:
I can add it in the config, But I find the constructor parameter is more explicit, because it is a needed config. As we discussed it looks like ApplicationRunner concept is going to be changed in future samza releases when that happens this is going to go away.

Reply:
Ideally, computeSamzaConfigs() should be invoked in the SamzaSqlApplicaiton, and user-config which runner is instantiated, w/o the need of a separate SamzaSqlApplicationRunner. We can do that in the next round of refactoring the APIs.